torch/_inductor/codegen/cpp_bmm_template.py

```diff
@@ -0,0 +1,171 @@
+{
+    const int64_t B = {{kernel.size(BY, -3, unwrapped=True)}};
+    {%- if num_threads > 1 %}
+    constexpr int64_t num_threads = {{num_threads}};
+    int64_t B_single_thread_block = (B / num_threads) * num_threads;
+
+    #pragma omp parallel for num_threads({{num_threads}})
+    {%- else %}
+    int64_t B_single_thread_block = B;
+    {%- endif %}
+    for (int64_t b_start = 0; b_start < B_single_thread_block; ++b_start) {
+        single_thread_mm(
+            &{{kernel.index(BX, ["b_start", 0, 0])}},
+            &{{kernel.index(BW, ["b_start", 0, 0])}},
+            &{{kernel.index(BY, ["b_start", 0, 0])}}
+            {%- if is_dynamic_M %},
+            {{kernel.size(GemmOut, -2)}}
+            {%- endif %}
+        );
+    }
+    for (int64_t b_start = B_single_thread_block; b_start < B; ++b_start) {
+        blocked_mm(
+            &{{kernel.index(BX, ["b_start", 0, 0])}},
+            &{{kernel.index(BW, ["b_start", 0, 0])}},
+            &{{kernel.index(BY, ["b_start", 0, 0])}}
+            {%- if is_dynamic_M %},
+            {{kernel.size(GemmOut, -2)}}
+            {%- endif %}
+        );
+    }
+}
+"""
+
+
+class CppBmmTemplate(CppPackedGemmTemplate):
+    def __init__(
+        self,
+        input_nodes,
+        layout: ir.Layout,
+        num_threads: int,
+        register_blocking: GemmBlocking,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+        name="bmm",
+    ):
+        super().__init__(
+            input_nodes,
+            layout,
+            num_threads,
+            register_blocking,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            epilogue_creator=epilogue_creator,
+            name=name,
+        )
+        self.should_pack_weights = False
+
+    @staticmethod
+    def add_choices(
+        choices,
+        layout,
+        input_nodes,
+        beta=1,
+        alpha=1,
+        has_bias=False,
+        trans_w=False,
+        input_indices=None,
+        should_pack_weights=False,
+        epilogue_creator: Optional[Callable[[ir.Buffer], ir.Pointwise]] = None,
+    ):
+        options = super(CppBmmTemplate, CppBmmTemplate)._get_params_for_choices(
+            layout=layout,
+            input_nodes=input_nodes,
+            beta=beta,
+            alpha=alpha,
+            has_bias=has_bias,
+            trans_w=trans_w,
+            input_indices=input_indices,
+            should_pack_weights=should_pack_weights,
+            epilogue_creator=epilogue_creator,
+        )
+        template = DataProcessorTemplateWrapper(CppBmmTemplate, **options)
+        template.maybe_append_choice(choices)
+        return template
+
+    def _get_default_reindexers(self, epilogue_nodes):
+        def reindexer(args):
+            if len(epilogue_nodes) == 0:
+                return args
+            return [0] + args
```

Comment:
Is there any special case we need to `reindexer` with a [0] inserted?

Reply:
Yes, when epilogue nodes exist they have 3D ranges but only 2D indexes (in `lowering.py:439`) because of taking the 2D slice to generate the sub-GEMM call. So we add a 0 index so we have 3D indexes and 3D ranges.

Reply:
This is worth a comment in the code.